include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,242 @@
+	std::map<std::string, ParameterValidityCallback> warn_if_callbacks_;
+	std::map<std::string, ParameterValidityCallback> error_if_callbacks_;
+	std::map<std::string, ParameterChangedCallback> on_changed_callbacks_;
+
+	// Store messages of failed validity callbacks
+	std::vector<std::string> errors_;
+	std::vector<std::string> warnings_;
+}
+
+// namespace Parameter
+using NodeParameter = Parameter<rclcpp::Parameter>;
+std::map<std::string, std::shared_ptr<NodeParameter>>
+NodeParameter::declare() override{ /* returns shared_from_this */ };
+
+/*
+ * Maintains all Parameter instances and does the heavy lifting with actually declaring and interfacing with dynamic
+ * rclcpp Parameters. Provides additional convenience functions for validating all parameters, printing, etc...
+ */
+class NodeConfig
+{
+public:
+	NodeConfig(const rclcpp::Node::SharedPtr& node, const std::string base_name = "");
+
+protected:
+
+	/*
+	 * Declare Parameter types, internally calls Parameter<ParameterT>::declare() to get
+	 * all actual node parameters which are then declared using the node parameter interface.
+	 */
+	template <typename ParameterT>
+	Parameter<ParameterT>& declare(Parameter<ParameterT>& parameter);
+
+	/*
+	 * Undeclare a parameter, remove all callbacks that have been registered using the declare() call
+	 */
+	template <typename ParameterT>
+	bool undeclare(const Parameter<ParameterT>& parameter);
+	bool undeclare(const std::string& name);
+
+	/*
+	 * Return the parameter for a given name, if not found parameter is nullptr
+	 */
+  // Not possible when working with references, maybe only for actual node parameters?
+	// std::shared_ptr<Parameter<ParameterT>> get(const std::string& name);
+	// std::shared_ptr<const Parameter<ParameterT>> get(const std::string& name) const noexcept;
+
+	//
+	// TODO: insert utility functions:  validate(), print, ostream<<, errors(), warnings()...
+	//
+
+private:
+	const rclcpp::Node::SharedPtr node_;
+	const std::string base_name_;
+
+  // Class similar to the one in rosparam_shortcuts/node_parameters.h
+  // implements ROS calls/checks/callbacks for node_parameters_
+  NodeParametersInterface node_parameters_interface_;
+
+  // Store parameters, only node parameters (rclcpp::Parameters) are stored inside NodeConfig
+	std::map<std::string, Parameter&> parameters_;
+	std::map<std::string, std::shared_ptr<NodeParameter>> node_parameters_;
+
+	// Needed? keep track of registered callbacks that are required for linking SetParameters requests or
+	// exposing high-level update changes (invalid/valid states)
+	// std::string<std::string, std::string> registered_node_parameter_callbacks_ids_;
+	// std::string<std::string, std::string> registered_parameter_callbacks_ids_;
+}
+
+struct ExampleConfig : public NodeConfig
+{
+  ExampleConfig(const rclcpp::Node::SharedPtr& node)
+    : NodeConfig(node)
+  {
+    declare(double_param)
+      .describe("This is a very important double parameter", "should be greater than 0 and not 7")
```

Comment:
hmm... this would only really work if the node parameters are actually declared after all these calls...

Reply:
Node parameters are loaded by the launch system after execution of the node starts once it provides the `set_parameter` interface which is started when you spin your ros node.  Is that what you are talking about?

Reply:
I mean that the `declare()` call will declare the node parameters which will already trigger all validity callbacks which have not been initialized at this point. Ideally, the builder functions `describe()`, `warnIf()`, etc are called before the parameter is actually declared. That means, either we need to simply run these before calling `declare()` which doesn't look as pretty or the actual declare calls of the node parameters is deferred to a `NodeConfig::declareAll()` function that is called after registering all the parameters.